### PR TITLE
Provide a dev mode for logging where filepath and line number are add…

### PR DESCRIPTION
**Pull request by wxtim**
…ed to the log message

This is a small change with no associated Issue.

On a number of occasions I have found it useful to add extra information about the source of logged messages to the log messages. I've done this manually for a number of tickets, and accidentally got a comment on how to do this accepted into master. On this basis I've added the ability to do this to the logging. 

Running 
```
cylc command -vvv
cylc command --debug -v
```
Will tell you where the code raising the log is.
Is this a high enough priority to avoid legit user use cases, or might we want `-vvvv`?

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required - Developer only feature.
- [x] No documentation update required.

### CODE REVIEW

File: tests/unit/test_loggingutil.py [hunk 1]
```diff
@@ -82,3 +85,27 @@ def test_value_error_raises_system_exit(
...
+            False,
+            '%(asctime)s %(levelname)-2s - %(message)s',
+            id='dev_info=False'
+        )
+    ]
+)
+def test_CylcLogFormatter__init__dev_info(dev_info, expect):
+    """dev_info switch changes the logging format string."""
+    formatter = CylcLogFormatter(dev_info=dev_info)
+    assert formatter._fmt == expect
```
Comment (kinow) [inline tests/unit/test_loggingutil.py:111]: Missing newline.

**Review by kinow — APPROVED** [review]
CI happy, code looks good.

- 1 v:

```bash
rm -rf ~/cylc-run/five/.service/ && cylc play --no-detach five -v
...
2021-07-22T20:55:44+12:00 INFO - Cold start from 1
2021-07-22T20:55:44+12:00 INFO - Scheduler: url=tcp://ranma:43088/ pid=146453
2021-07-22T20:55:44+12:00 INFO - Workflow publisher: url=tcp://ranma:43079
...
```

- 2 v:

```bash
rm -rf ~/cylc-run/five/.service/ && cylc play --no-detach five -vv
...
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five: directory created
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five/log/workflow: directory created
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five/log/job: directory created
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five/log/flow-config: directory created
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five/share: directory created
2021-07-22T20:57:06+12:00 DEBUG - /home/kinow/cylc-run/five/work: directory created
2021-07-22T20:57:06+12:00 DEBUG - Extracting etc/job.sh to /home/kinow/cylc-run/five/.service/etc/job.sh
2021-07-22T20:57:06+12:00 DEBUG - Starting
2021-07-22T20:57:06+12:00 DEBUG - auth received API command b'CURVE'
2021-07-22T20:57:06+12:00 DEBUG - Configure curve: *[/home/kinow/cylc-run/five/.service/client_public_keys]
2021-07-22T20:57:06+12:00 DEBUG - Reading file /home/kinow/cylc-run/five/flow.cylc
2021-07-22T20:57:06+12:00 DEBUG - Processed configuration dumped: /home/kinow/cylc-run/five/flow.cylc.processed
2021-07-22T20:57:06+12:00 DEBUG - Expanding [runtime] namespace lists and parameters
2021-07-22T20:57:06+12:00 DEBUG - Parsing the runtime namespace hierarchy
2021-07-22T20:57:06+12:00 DEBUG - Parsing [special tasks]
2021-07-22T20:57:06+12:00 DEBUG - Parsing the dependency graph
2021-07-22T20:57:06+12:00 INFO - Cold start from 1
2021-07-22T20:57:06+12:00 DEBUG - Loaded main loop plugin "health check
	": * health_check
2021-07-22T20:57:06+12:00 DEBUG - Loaded main loop plugin "prune flow labels
	": * prune_flow_labels
2021-07-22T20:57:06+12:00 INFO - Scheduler: url=tcp://ranma:43032/ pid=146831
2021-07-22T20:57:06+12:00 INFO - Workflow publisher: url=tcp://ranma:43014
...
```

- 3 v:

```bash
rm -rf ~/cylc-run/five/.service/ && cylc play --no-detach five -vvv
...
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five/log/workflow: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five/log/job: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five/log/flow-config: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five/share: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/pathutil.py:130]
	2021-07-22T20:58:22+12:00 DEBUG - /home/kinow/cylc-run/five/work: directory created
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/resources.py:59]
	2021-07-22T20:58:22+12:00 DEBUG - Extracting etc/job.sh to /home/kinow/cylc-run/five/.service/etc/job.sh
[/home/kinow/Development/python/workspace/cylc-flow/venv/lib/python3.8/site-packages/zmq/auth/base.py:53]
	2021-07-22T20:58:22+12:00 DEBUG - Starting
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/parsec/fileparse.py:298]
	2021-07-22T20:58:22+12:00 DEBUG - Reading file /home/kinow/cylc-run/five/flow.cylc
[/home/kinow/Development/python/workspace/cylc-flow/venv/lib/python3.8/site-packages/zmq/auth/thread.py:93]
	2021-07-22T20:58:22+12:00 DEBUG - auth received API command b'CURVE'
[/home/kinow/Development/python/workspace/cylc-flow/venv/lib/python3.8/site-packages/zmq/auth/base.py:114]
	2021-07-22T20:58:22+12:00 DEBUG - Configure curve: *[/home/kinow/cylc-run/five/.service/client_public_keys]
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/parsec/fileparse.py:470]
	2021-07-22T20:58:22+12:00 DEBUG - Processed configuration dumped: /home/kinow/cylc-run/five/flow.cylc.processed
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/config.py:291]
	2021-07-22T20:58:22+12:00 DEBUG - Expanding [runtime] namespace lists and parameters
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/config.py:1044]
	2021-07-22T20:58:22+12:00 DEBUG - Parsing the runtime namespace hierarchy
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/config.py:367]
	2021-07-22T20:58:22+12:00 DEBUG - Parsing [special tasks]
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/config.py:1878]
	2021-07-22T20:58:22+12:00 DEBUG - Parsing the dependency graph
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/scheduler.py:639]
	2021-07-22T20:58:22+12:00 INFO - Cold start from 1
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/main_loop/__init__.py:349]
	2021-07-22T20:58:22+12:00 DEBUG - Loaded main loop plugin "health check
	": * health_check
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/main_loop/__init__.py:349]
	2021-07-22T20:58:22+12:00 DEBUG - Loaded main loop plugin "prune flow labels
	": * prune_flow_labels
[/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/scheduler.py:528]
	2021-07-22T20:58:22+12:00 INFO - Scheduler: url=tcp://ranma:43099/ pid=147208
...
```

@wxtim I don't mind so much the extra verbose format, but it we use a single line in the output, it could be helpful when `grep`ing. Although we could use `grep -A2` when searching for a specific file name, it would still be easier if it were in the same line for others I guess.

Other than that, looks perfect, no more comments :tada: !

Bruno

**Comment by wxtim** [comment]
@kinow  removed '\n' from message.

File: cylc/flow/loggingutil.py [hunk 2]
```diff
@@ -58,17 +58,29 @@ class CylcLogFormatter(logging.Formatter):
...
-            '%(asctime)s %(levelname)-2s - %(message)s',
-            '%Y-%m-%dT%H:%M:%S%Z')
+        if dev_info is True:
+            # Add location message created in codebase when in very very, very
+            # verbose mode.
+            logging.Formatter.__init__(
+                self,
+                (
+                    '[%(pathname)s:%(lineno)d] '
+                    '%(asctime)s %(levelname)-2s - %(message)s'
```
Comment (oliver-sanders) [inline cylc/flow/loggingutil.py:76]: Could we have the timestamp first (always the same length).

Comment (oliver-sanders) [reply cylc/flow/loggingutil.py:76]: Also, worth checking that this doesn't conflict with the no timestamp mode.

Comment (wxtim) [reply cylc/flow/loggingutil.py]: Having the timestamp first fixed the conflict which _did_ exist with the `--no-timestamp` option.

**Comment by oliver-sanders** [comment]
Tested with `-vvv` and `--no-timestamp`.

The full absolute path to the file is a bit long and not so good for reporting issues, could I suggest using `module` rather than `pathname`. It's not the full module name (i.e. `cylc.flow.whatever`) but it's good enough to work out where the message came from.

Suggestion:

```diff
diff --git a/cylc/flow/loggingutil.py b/cylc/flow/loggingutil.py
index 6899e7ca0..290d1d57b 100644
--- a/cylc/flow/loggingutil.py
+++ b/cylc/flow/loggingutil.py
@@ -66,21 +66,18 @@ class CylcLogFormatter(logging.Formatter):
         self.max_width = self.MAX_WIDTH
         self.wrapper = None
         self.configure(timestamp, color, max_width)
+        prefix = '%(asctime)s %(levelname)-2s'
+
         if dev_info is True:
             # Add location message created in codebase when in very very, very
             # verbose mode.
-            logging.Formatter.__init__(
-                self,
-                (
-                    '%(asctime)s %(levelname)-2s - [%(pathname)s:%(lineno)d] '
-                    '%(message)s'
-                ),
-                '%Y-%m-%dT%H:%M:%S%Z')
-        else:
-            logging.Formatter.__init__(
-                self,
-                '%(asctime)s %(levelname)-2s - %(message)s',
-                '%Y-%m-%dT%H:%M:%S%Z')
+            prefix += ' [%(module)s:%(lineno)d]'
+
+        logging.Formatter.__init__(
+            self,
+            prefix + ' - %(message)s',
+            '%Y-%m-%dT%H:%M:%S%Z'
+        )
 ```

Example:

```
2021-08-09T11:07:18+01:00 WARNING [workflow_files:1443] - The filename
	"suite.rc" is deprecated in favour of "flow.cylc"
```

File: cylc/flow/loggingutil.py [hunk 3]
```diff
@@ -105,6 +117,23 @@ def formatTime(self, record, datefmt=None):
         return get_time_string_from_unix_time(record.created)
 
 
+class CylcLogFormatterPlus(CylcLogFormatter):
+    def __init__(self, timestamp=True, color=False, max_width=None):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/flow/loggingutil.py]: I have no idea. Claiming that I've been away for a week. It certainly appears to be surplus to requirements: Deleted.

File: tests/unit/test_loggingutil.py [hunk 4]
```diff
@@ -82,3 +85,27 @@ def test_value_error_raises_system_exit(
...
+            id='dev_info=True'
+        ),
+        param(
+            False,
+            '%(asctime)s %(levelname)-2s - %(message)s',
+            id='dev_info=False'
+        )
+    ]
+)
+def test_CylcLogFormatter__init__dev_info(dev_info, expect):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply tests/unit/test_loggingutil.py:108]: On consideration I think that I probably agree with you that this is insufficient, although as a unit test for a fairly trivial function it make sense for the unit test to be a bit trivial, so I think that still has purpose.

**Review by oliver-sanders — APPROVED** [review]

**Review by MetRonnie — APPROVED** [review]

**Pull request merged**